### ISSUE TITLE
Mutation is lost when network connection is bad

### ISSUE THREAD

**Issue by nourahassan** [awslabs/aws-mobile-appsync-sdk-ios#189]
I am using appsync in my iOS project and it can successfully save offline objects when network is on. But if I am having a case where I have dynamoDB object with s3 file, after saving the file successfully to s3, the network becomes bad and the dynamoDB is not saved even when a good connection is back.
  
**To Reproduce**
1. Configure a mutation with s3 file input
2. Add breakpoint to uploadS3 completion block at the extension `extension AWSS3TransferUtility: AWSS3ObjectManager`
3. Create your object and save when it stops at the breakpoint, restart your router (this is how I produce network error)
4. See error about `Did not receive a successful HTTP code`
5. After connection is back, the object will not saved to DynamoDB

**Expected behavior**
It is expected that it will not remove the mutation from the queue unless it is saved successfully. So it should retry saving after connection is good.

**Environment(please complete the following information):**
 - AppSync SDK Version: 2.10.2
 - Dependency Manager: Carthage
 - Swift Version : 4.2

**Device Information (please complete the following information):**
 - Device: iPhone6s
 - iOS Version: iOS 12.1.4

FYI, Android had a [similar issue](https://github.com/awslabs/aws-mobile-appsync-sdk-android/issues/82),  that is fixed now.

**Comment by palpatim** [comment]
Hi @nourahassan,

Thanks for the bug report. I'll investigate and update this thread once I have more info.

**Comment by nourahassan** [comment]
any updated for this ticket? It happened a lot with poor connection or when device changes between wifi to cellular. 

I think the problem is for Reachability it thinks there is a connection like while router restarts but AppSync should handle the timeout error as offline and retry again when reachability changed.

**Comment by palpatim** [comment]
Hi @nourahassan,

In our [2.11.0](https://github.com/awslabs/aws-mobile-appsync-sdk-ios/releases/tag/2.11.0) release, we did quite a bit of work around more aggressively retrying mutations (and explicitly not erroring them out) in bad network situations. Are you able to upgrade to ~>2.11.0 to try?

Please note, if you're building in Xcode 10.2 or greater, you need to be using ~> 2.12.0 -- the older version of SQLite.swift has an issue that causes persistent cache failures. Although we didn't explicitly look at the mutation queue with that configuration, it's possible that could also be causing an issue.

**Comment by nourahassan** [comment]
We made a release at the weekend after doing `carthage update`,  Xcode `10.2.1` but the version was `2.11.0` not sure why?!

We are making another one today will check logs in next two days and let you know if the issue is fixed or not.

**Comment by nourahassan** [comment]
@palpatim I am not seeing the **timed out exception** since the update to **2.12**, instead I can see errors about `Software caused connection abort` or `(403 forbidden) Did not receive a successful HTTP code`

Mutations with the first error are saved to server later which is great!
Mutations with `403` error, are not saved to server. Probably it failed to refresh token because mutations before and after which call same API are saved correctly.

**Comment by palpatim** [comment]
@nourahassan

We'll take a look at this, but in the meantime, can you confirm: are you using AWSMobileClient to manage automatic refreshes? I need to refresh my memory on our auto-retry code, but my recollection is that 403 isn't automatically retried, so we may have a bug where a mutation gets a 403, MobileClient auto-refreshes the token, and subsequent mutations succeed. But understanding your call pattern would be helpful. Thanks!

**Comment by nourahassan** [comment]
Yes, I am not refreshing token manually. As you said when MobileClient refreshed it, next mutations passed.

**Comment by palpatim** [comment]
Thanks for clarifying, we're investigating.

**Comment by palpatim** [comment]
@nourahassan 

I'm not immediately seeing issues in the code, although I haven't been able to replicate the 403 condition locally. As I work on tests to verify this, can you confirm whether the 403 issue happens only on mutations with S3 objects? That is, if you encounter a 403 on a mutation with no complex object, does the mutation appear on the server as expected? Or are you seeing this behavior on mutations both with and without complex objects?

**Comment by nourahassan** [comment]
It doesn't happen on mutation with s3 object, it is just a mutation to save dynamoDB object. The mutation with the given error is not showing on the db. I couldn't see entry on cloud watch logs.

As workaround for the initial bug I separated s3 object to another mutations that only save file to s3, didn't back to complex mutation yet.
